DexPriceAggregatorUniswapV3/contracts/DexPriceAggregatorUniswapV3.sol

```diff
@@ -0,0 +1,409 @@
+    /// @param _twapPeriod Number of seconds in the past to consider for the TWAP rate
+    /// @return amountOut Amount of tokenOut received for ethAmountIn of ETH
+    function ethToAsset(
+        uint256 _ethAmountIn,
+        address _tokenOut,
+        uint256 _twapPeriod
+    ) public view returns (uint256 amountOut) {
+        address tokenIn = weth;
+        address pool = _getPoolForRoute(
+            PoolAddress.getPoolKey(tokenIn, _tokenOut, defaultPoolFee)
+        );
+        return
+            _fetchAmountFromSinglePool(
+                tokenIn,
+                _ethAmountIn,
+                _tokenOut,
+                pool,
+                _twapPeriod
+            );
+    }
+
+    /************************
+     * Management functions *
+     ************************/
+
+    /// @notice Override the Uniswap V3 pool queried on a tokenA:tokenB route
+    /// @dev Override can be reset by using address(0) for _pool
+    /// @param _tokenA Address of an ERC20 token contract
+    /// @param _tokenB Address of another ERC20 token contract
+    /// @param _pool Address of a Uniswap V3 pool constructed with _tokenA and _tokenB
+    function setPoolForRoute(
+        address _tokenA,
+        address _tokenB,
+        address _pool
+    ) public onlyOwner {
+        PoolAddress.PoolKey memory poolKey = PoolAddress.getPoolKey(
+            _tokenA,
+            _tokenB,
+            uint24(0) // pool fee is unused
+        );
+        if (_pool != address(0)) {
+            require(
+                poolKey.token0 == IUniswapV3Pool(_pool).token0() &&
+                    poolKey.token1 == IUniswapV3Pool(_pool).token1(),
+                "Tokens or pool not correct"
+            );
+        }
+        overriddenPoolForRoute[_identifyRouteFromPoolKey(poolKey)] = _pool;
+        emit PoolForRouteSet(poolKey.token0, poolKey.token1, _pool);
+    }
+
+    /// @notice Fetch the Uniswap V3 pool to be queried for a tokenA:tokenB route
+    /// @param _tokenA Address of an ERC20 token contract
+    /// @param _tokenB Address of another ERC20 token contract
+    /// @return pool Address of a Uniswap V3 pool constructed with _tokenA and _tokenB
+    function getPoolForRoute(address _tokenA, address _tokenB)
+        public
+        view
+        returns (address pool)
+    {
+        return
+            _getPoolForRoute(
+                PoolAddress.getPoolKey(_tokenA, _tokenB, defaultPoolFee)
+            );
+    }
+
+    /**************************
+     * Utility view functions *
+     **************************/
+
+    /// @notice Fetch a Uniswap V3 pool's current "spot" and TWAP tick values
+    /// @param _pool Address of a Uniswap V3 pool
+    /// @param _twapPeriod Number of seconds in the past to consider for the TWAP rate
+    /// @return spotTick The pool's current "spot" tick
+    /// @return twapTick The twap tick for the last _twapPeriod seconds
+    function fetchCurrentTicks(address _pool, uint32 _twapPeriod)
```

Comment:
_twapPeriod should be greater than 0

libraries/OracleLibrary.sol

function consult(address pool, uint32 period)
internal
view
returns (int24 timeWeightedAverageTick)
{
require(period != 0, "BP"); ///// -----------> this